fw/mmap_buffer.c

```diff
@@ -0,0 +1,329 @@
+static TfwMmapBufferHolder *holders[MAX_HOLDERS];
+static int holders_cnt;
+
+static int dev_file_open(struct inode *ino, struct file *filp);
+static int dev_file_close(struct inode *ino, struct file *filp);
+static int dev_file_mmap(struct file *filp, struct vm_area_struct *vma);
+static void dev_file_vm_close(struct vm_area_struct *vma);
+
+static const struct file_operations dev_fops = {
+	.open  = dev_file_open,
+	.release  = dev_file_close,
+	.mmap  = dev_file_mmap,
+};
+
+static const struct vm_operations_struct dev_vm_ops = {
+	.close = dev_file_vm_close
+};
+
+static int
+dev_file_open(struct inode *ino, struct file *filp)
+{
+	TfwMmapBufferHolder *holder;
+	int i;
+
+	for (i = 0; i < holders_cnt; ++i) {
+		if (strcmp(holders[i]->dev_name,
+			   (char *)filp->f_path.dentry->d_iname) == 0) {
+			holder = holders[i];
+			goto found;
+		}
+	}
+
+	return -EINVAL;
+
+found:
+	if (atomic_read(&holder->is_freeing))
+		return -ENOENT;
+	if (atomic_read(&holder->dev_is_opened))
+		return -EBUSY;
+
+	atomic_set(&holder->dev_is_opened, 1);
+	filp->private_data = holder;
+
+	return 0;
+}
+
+static int
+dev_file_close(struct inode *ino, struct file *filp)
+{
+	TfwMmapBufferHolder *holder = filp->private_data;
+
+	atomic_set(&holder->dev_is_opened, 0);
+	return 0;
+}
+
+/*
+ * This function handles the mapping of ring buffers into user space. Each
+ * buffer should be mapped by user space with an offset calculated as
+ * full_buffer_size * cpu_num, where full_buffer_size is the size of buffer data
+ * plus TFW_MMAP_BUFFER_DATA_OFFSET; cpu_num is a number of CPU in a row. This
+ * allows determining which CPU's buffer should be mapped based on the offset.
+ */
+static int
+dev_file_mmap(struct file *filp, struct vm_area_struct *vma)
+{
+	TfwMmapBufferHolder *holder = filp->private_data;
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+	unsigned long pfn, size, area_size, user_addr;
+	int cpu_num, cpu_id, i;
+
+	size = vma->vm_end - vma->vm_start;
+
+#define FULL_SIZE TFW_MMAP_BUFFER_FULL_SIZE(buf->size)
+
+	if (size == TFW_MMAP_BUFFER_DATA_OFFSET)
+		area_size = size;
+	else if (size == FULL_SIZE)
+		area_size = FULL_SIZE;
+	else
+		return -EINVAL;
+
+#define NTH_ONLINE_CPU(n) ({		\
+	int cpu, res = -1, i = 0;	\
+	for_each_online_cpu(cpu) {	\
+		if (i == n) {		\
+			res = cpu;	\
+			break;		\
+		}			\
+		++i;			\
+	}				\
+	res;				\
+})
+
+	cpu_num = vma->vm_pgoff / (area_size / PAGE_SIZE);
+	cpu_id = NTH_ONLINE_CPU(cpu_num);
+	if (cpu_id < 0)
+		return -EINVAL;
+
+	buf = *per_cpu_ptr(holder->buf, cpu_id);
+
+	pfn = page_to_pfn(holder->mem[cpu_id].buf_page);
+	if (remap_pfn_range(vma, vma->vm_start, pfn,
+			    TFW_MMAP_BUFFER_DATA_OFFSET, vma->vm_page_prot))
+		return -EAGAIN;
+
+	if (area_size == FULL_SIZE) { /* entire buffer is mapping */
+		pfn = page_to_pfn(holder->mem[cpu_id].data_page);
+		user_addr = vma->vm_start + TFW_MMAP_BUFFER_DATA_OFFSET;
+		for (i = 0; i < 2; ++i) {
+			if (remap_pfn_range(vma, user_addr, pfn, buf->size,
```

Comment:
who unmap previously mmaped area if this call fails?

Reply:
If this call fails user space gets MAP_FAILED and exits with the error "Failed to map buffer".
As I understand, all the mmaped to user space areas unmap in the process of closing.